Daml: Add MintingDelegation support

dfordivam opened the pull request:
This PR includes the feedback from https://github.com/hyperledger-labs/splice/pull/1627

The implementation has been modified as the `dso` is no longer an observer on these.
The `Expire` APIs are no longer needed and have been removed.

Fixes #3356 


### Pull Request Checklist

#### Cluster Testing
- [ ] If a cluster test is required, comment `/cluster_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.
- [ ] If a hard-migration test is required (from the latest release), comment `/hdm_test` on this PR to request it, and ping someone with access to the DA-internal system to approve it.

#### PR Guidelines
- [ ] Include any change that might be observable by our partners or affect their deployment in the [release notes](https://github.com/hyperledger-labs/splice/blob/main/docs/src/release_notes.rst).
- [ ] Specify fixed issues with `Fixes #n`, and mention issues worked on using `#n`
- [ ] Include a screenshot for frontend-related PRs - see [README](https://github.com/hyperledger-labs/splice/blob/main/TESTING.md#running-and-debugging-integration-tests) or use your favorite screenshot tool


#### Merge Guidelines
- [ ] Make the git commit message look sensible when squash-merging on GitHub (most likely: just copy your PR description).

meiersi-da commented on daml/splice-wallet/daml/Splice/Wallet/MintingDelegation.daml:
```diff
@@ -0,0 +1,149 @@
+      do pure MintingDelegation_WithdrawResult with dummyUnitField = ()
+
+    -- We do not charge a fee for MintingDelegation but we make them expire
+    -- and force paying traffic to renew them to protect against attacks.
```
We can remove this choice now that the `DSO` is no longer involved in the delegation.

meiersi-da commented on token-standard/splice-token-standard-test/daml/Splice/Testing/Registries/AmuletRegistry/Parameters.daml:
```diff
@@ -88,6 +88,8 @@ defaultAmuletConfig = AmuletConfig with
   -- Amount of the AppRewardCoupon contract that a FeaturedAppActivityMarker is converted to.
   featuredAppActivityMarkerAmount = Some defaultFeaturedAppActivityMarkerAmount
 
+  mintingDelegationLifetime = None
```
let's remove this param

dfordivam commented on daml/splice-wallet/daml/Splice/Wallet/MintingDelegation.daml:
```diff
@@ -0,0 +1,149 @@
+           dso
+           beneficiary
+           delegate
+           expiresAt = now `addRelTime` getMintingDelegationLifetime config
```
Right, this was a remnant from the sketch PR, and I have removed. Now with the DSO not involved, the `expiresAt` has a different meaning from what was originally designed for.

dfordivam commented:
Ok, I have made this simpler and now this is limited to `splice-wallet` only.

meiersi-da commented on daml/splice-wallet/daml/Splice/Wallet/MintingDelegation.daml:
```diff
@@ -0,0 +1,115 @@
+defaultAmuletMergeLimit : Int
+defaultAmuletMergeLimit = 10
+
+template MintingDelegation
```
add comment explaining what this template is intended to do

meiersi-da reviewed (approved):
Thanks. Looks good up to the missing comments.

dfordivam commented:
I have added `FeaturedAppActivityMarker` support for the delegate taking inspiration (and helper APIs) from `Splice.Util.Token.Wallet.MergeDelegation`. And added comments.

meiersi-da commented on daml/splice-wallet/daml/Splice/Wallet/MintingDelegation.daml:
```diff
@@ -0,0 +1,132 @@
+        openRoundCid : ContractId OpenMiningRound
+          -- ^ The open mining round for which to record activity.
+        optFeaturedAppRight : Optional FeaturedAppRightCall
+          -- ^ The featured app right to use to feature the delegate's activity.
```
Upon further consideration, I think this goes in the wrong direction. The direction of "Minting rewards for minting rewards" seems problematic.

More generally, the contracts here implement purely operational tools, and thus do not qualify for app rewards. I'd thus suggest to drop the app reward support here.

This also makes it consistent with the choice below.

dfordivam replied:
@meiersi-da I have removed the `FeaturedAppRight` related code. The PR is ready for review

meiersi-da reviewed (changes requested):
Thanks for the update. Let's make sure we are aligned on the app rewards topic before we merge this.

meiersi-da reviewed (approved):
Nice work! Thanks a lot!
